File: examples/benchmarks/1_scattering_kernel_speed/compute_kernel.jl
using Pkg
Pkg.activate(dirname(@__DIR__))
Pkg.instantiate()

using CSV
using BenchmarkTools
using DataFrames
using ProgressBars
using ScatteringOptics
using StableRNGs
using VLBISkyModels

@info "This benchmark script will compare the time taken to compute the visibility function of the scatteirng kernel between eht-imaging and ScatteringOptics.jl"

# number of uvsamples
const nsamples = Int64.(floor.(exp10.(range(1, 6, length=20))))

# maximum uvd
const uvdmax = Float64(10^10)

# reference frequency
const λcm = 0.13
const νref = ScatteringOptics.λcm2ν(λcm)

@info "Benchmarking ScatteringOptics.jl"

# creating the scattering models
const sms = [DipoleScatteringModel(), vonMisesScatteringModel(), PeriodicBoxCarScatteringModel()]
const kernels = map(sm -> kernelmodel(sm; νref=νref), sms)

# a function to compute visibility
@inline function so_compute!(vis, sm, λcm, uvec, vvec)
    for i in 1:length(uvec)
        @inbounds vis[i] = visibility_point_approx(sm, λcm, uvec[i], vvec[i])
    end
    return nothing
end

# Benchmarking
so_results=Array{Float64,2}(undef, 3, length(nsamples))
for i in ProgressBar(1:length(nsamples))
    local nsample = nsamples[i]
    local stablerng = StableRNG(nsample)
    local uvec = rand(stablerng, nsample) .* 1e10
    local vvec = rand(stablerng, nsample) .* 1e10
    #local p = UnstructuredDomain((U=uvec, V=vvec))
    local vis = Vector{Float64}(undef, nsample)
    for j in 1:3
        local sm = sms[j]
        local kernel = kernels[j]
        #local out = @benchmark visibilitymap!($vis, $kernel, $p) samples=100
        local out = @benchmark so_compute!($vis, $sm, $λcm, $uvec, $vvec) samples=100
        @inbounds so_results[j, i] = minimum(out).time/1e9
    end
end

df = DataFrame(
    samples = nsamples,
    dipole = so_results[1, :],
    vonmises = so_results[2, :],
    boxcar = so_results[3, :]
)
CSV.write("so_compute_kernel.csv", df)

File: examples/benchmarks/1_scattering_kernel_speed/compute_kernel.py
import ehtim as eh
import numpy as np
import pandas as pd
import timeit
import tqdm
import random

print("This script will benchmark eht-imaging's computation of scattering kernel.")

# numbe of samples
nsamples = np.int64(np.floor(np.logspace(1, 6, 20)))

# scattering models
sm_dp = eh.scattering.ScatteringModel(model="dipole")
sm_vm = eh.scattering.ScatteringModel(model="von_Mises")
sm_bc = eh.scattering.ScatteringModel(model="boxcar")

dpset = []
vmset = []
bcset = []
for nsample in tqdm.tqdm(nsamples):
    # fix the seed for the reproducibility
    np.random.seed(nsample)

    # sample uv points
    u = np.random.uniform(-1e10, 1e10, nsample)
    v = np.random.uniform(-1e10, 1e10, nsample)

    def compute_dp():
        return sm_dp.Ensemble_Average_Kernel_Visibility(u=u, v=v, wavelength_cm=0.13)

    def compute_vm():
        return sm_vm.Ensemble_Average_Kernel_Visibility(u=u, v=v, wavelength_cm=0.13)

    def compute_bc():
        return sm_bc.Ensemble_Average_Kernel_Visibility(u=u, v=v, wavelength_cm=0.13)

    dpt = np.median(timeit.repeat(compute_dp, repeat=100, number=1))
    vmt = np.median(timeit.repeat(compute_vm, repeat=100, number=1))
    bct = np.median(timeit.repeat(compute_bc, repeat=100, number=1))
    dpset.append(dpt)
    vmset.append(vmt)
    bcset.append(bct)

df = pd.DataFrame()
df["samples"] = nsamples
df["dipole"] = dpset
df["vonmises"] = vmset
df["boxcar"] = bcset
df.to_csv("eh_compute_kernel.csv")
